src/Abuse/Adapters/Redis.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace Utopia\Abuse\Adapters;
+
+use Redis as Client;
+
+class Redis extends TimeLimit
+{
+    public const NAMESPACE = 'abuse';
+
+    /**
+     * @var Client
+     */
+    protected Client $redis;
+
+    public function __construct(string $key, int $limit, int $seconds, Client $redis)
+    {
+        $this->redis = $redis;
+        $this->key = $key;
+        $time = (int) \date('U', (int) (\floor(\time() / $seconds)) * $seconds);
+        $this->time = strval($time);
+        $this->limit = $limit;
+    }
+
+    /**
+     * Undocumented function
+     *
+     * @param string $key
+     * @param string $datetime
+     * @return integer
+     */
+    protected function count(string $key, string $datetime): int
+    {
+        if (0 == $this->limit) { // No limit no point for counting
+            return 0;
+        }
+
+        if (! \is_null($this->count)) { // Get fetched result
+            return $this->count;
+        }
+
+        /** @var string $count */
+        $count = $this->redis->get(self::NAMESPACE . '__'. $key .'__'. $datetime);
+        if (!$count) {
+            $this->count = 0;
+        } else {
+            $this->count = intval($count);
+        }
+
+        return $this->count;
+    }
+
+    /**
+     * @param  string  $key
+     * @param  string  $datetime
+     * @return void
+     *
+     */
+    protected function hit(string $key, string $datetime): void
+    {
+        if (0 == $this->limit) { // No limit no point for counting
+            return;
+        }
+
+        /** @var string $count */
+        $count = $this->redis->get(self::NAMESPACE . '__'. $key .'__'. $datetime);
+        if (!$count) {
+            $this->count = 0;
+        } else {
+            $this->count = intval($count);
+        }
+
+        $this->redis->incr(self::NAMESPACE . '__'. $key .'__'. $datetime);
+        $this->count++;
+    }
+
+    /**
+     * Get abuse logs
+     *
+     * Return logs with an offset and limit
+     *
+     * @param  int|null  $offset
+     * @param  int|null  $limit
+     * @return array<string, mixed>
+     */
+    public function getLogs(?int $offset = null, ?int $limit = 25): array
+    {
+        // TODO limit potential is SCAN but needs cursor no offset
+        $cursor = null;
+        $keys = $this->redis->scan($cursor, self::NAMESPACE . '__*', $limit);
```

Comment:
I guess using HASH in this adapter is not the case here.
So using SCAN with a cursor is the next best thing.
I think we are good as long as we avoid using KEYS *.
You can load it with kyes and then use SCAN to loop and inspect slow logs with redisinsight, just to make sure.

Reply:
I am not sure I understand what you mean?

Reply:
Tested with redis insight. And scan works as expected.